### DISCUSSION TITLE
[v5] Reference type does not give access to primary keys when paired with generics

### DISCUSSION THREAD

**Discussion by darkbasic** [mikro-orm/mikro-orm#2311] [Q&A]
**Describe the bug**
```ts
export class Session<CustomUser extends IUser<any, any, any>> {
  id!: number;

  user: IdentifiedReference<CustomUser>;

  [...]
}

interface MyUser {
  id: number;
  username?: string;
  services: Collection<any>;
  emails: Collection<any>;
  sessions: Collection<any>;
  deactivated: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const toSession1 = async (session: Session<MyUser> | null): Promise<ISession | null> =>
  session && {
    ...session,
    id: String(session.id),
    userId: String(session.user.id), //WORKS
    [...]
  };

const toSession2 = async <T extends MyUser>(session: Session<T> | null): Promise<ISession | null> =>
  session && {
    ...session,
    id: String(session.id),
    userId: String(session.user.id), //Property id does not exist on type Reference<T>
    [...]
  };
```

The Reference type does not give me synchronous access to the primary key (which should always be loaded) whenever I use a generic for the relevant type.
**Versions**

| Dependency | Version |
| - | - |
| node | v14.18.0 |
| typescript | 4.4.4 |
| mikro-orm | 5.0.0-dev.424 |

**Comment by B4nan** [comment]
I believe that is a TS limitation, saw very similar issues before too, the type used in extends is not respected. If you hover over it, you can see just `T` and it is no longer constrained. Without having there proper entity type, the inference cannot work.

![image](https://user-images.githubusercontent.com/615580/139146334-30dfcbb1-99a7-4fc0-8229-f6f1ef35086b.png)

**Comment by B4nan** [comment]
Not very nice, but you could do this to get around it:

```ts
export class Session<CustomUser extends IUser> {
  id!: number;
  user!: CustomUser extends IUser ? IdentifiedReference<CustomUser> : IdentifiedReference<CustomUser>;
}
```

[playground](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcDCEB2BnGUCuAxjNADwAqcAvHPpgNaYQDumAfNXJsM3ABQA6IQEMoAc2wAuOMMyIA2gF0AlNQ7kA3AChQkWAmSoAIgEtiJrKMQVOsxBxoBvOPPrTcUE5jGLplAL7autDwSChwAKIghAA2+AAmwABidOZY2BQANHAA0nCgMMCY8dhw9MCIEABmcOQOtfI5ivkghcWlKZhpmHAA-FzAAG7AUHDSOUHgIQbhaMK4WXAASvWUBUUly321Y8uTeqGGcACS2ACqDEysFKstbZt0jCw9-Y9XPevtO-14+KjS3GGowBQxG2jCqAAgnIIpgYCYkDdOAAFUTw4QxG77aYQuDIzwAWys+IgKFg1jqnDWrQ2pWcAH0TPFpHY4P4tHBtgByRnxLlwAA+cA8XjEu2p9zptHwTJZcjZHO5+BlfPFd1pcGcspk8vZnP6XKZ-JBQOx+li81KS2AVRGRUIwBsn020MQsPhiLqHEcisI6V+xGgfDAnkGwkK+ThCMQflUznZ7Nxx0SUaqJmA8Wttqg9sdEo1rvd0Zu2WReWdpXKlRqlCFb2eKMJxKgpJGnrY9WcjTgXnQ8xgpDL2Sr1VqbF8DSabLgADJlja7V082xtFok2dsCNOFrmVx8ASAEZg2ibqCYYQE4C9dx4UVs1fBc0xS1wADKwGw2AsmFIaHwuAgAkNy3CsTmAqBvUVJkAEIAX3I8oG0Tl-xGWD0H-EggNPdUvmOcDtmTIp4TTDMs0XB1fwwwDwI4aRCNTdNMwXHMl0ogCsJGFctHZLQvEKKAqmEB04AAWUQfCfU5bVMHgsFFRQs8LyvG9PG8JC4ESIT4TDQpdwPCAIBiYBZHUwgc3DDNIRgaQjAs9T8DAeILPiKybLs7itC0P0cFCCB30-b8AEZbGwRAun4TcAqwaR-K-LBSDEmjBT3GIYlUKg2EVSK4p6Gc50kzk4CEARsu-TJFSk3dX1vbw+FKrABCZZRysKk8RmTGKarEOqPxygQFMa+JmrgAB6EaAHUAHklhyV9FUCTzvNwBA-N678ACYQrCwg4CdGkvkS082B6qLMBitb4spIUZNS9LMs5ercvyiqiqER6WsK7VqtU7rHsG5qXoUjq3y6k6+oGprsjGkkySQHt4g0iAPy4CB4FAExlqwGZUDIliKLqebVyAA)
